base/multidimensional.jl

```diff
@@ -487,9 +487,8 @@ rcum_promote_type{T,S<:Number}(op, ::Type{T}, ::Type{S}) = promote_op(op, T, S)
 rcum_promote_type{T<:Number}(op, ::Type{T}) = rcum_promote_type(op, T,T)
 rcum_promote_type{T}(op, ::Type{T}) = T
 
-# handle sums of Vector{Bool} and similar.   it would be nice to handle
-# any AbstractArray here, but it's not clear how that would be possible
-rcum_promote_type{T,N}(op, ::Type{Array{T,N}}) = Array{rcum_promote_type(op,T), N}
+# handle sums of Vector{Bool} and similar.
+rcum_promote_type{T,N}(op, ::Type{<:AbstractArray{T,N}}) = Array{rcum_promote_type(op,T), N}
```

Comment:
should probably fall back to similar though, otherwise this might not be the right behavior for shared or distributed arrays